python/cuml/experimental/explainer/permutation_shap.pyx

```diff
@@ -0,0 +1,451 @@
+                                 <int> self.n_rows,
+                                 <int> n_cols,
+                                 <float*> row_ptr,
+                                 <int*> idx_ptr,
+                                 <bool> row_major)
+            else:
+                permutation_dataset(handle_[0],
+                                 <double*> masked_ptr,
+                                 <double*> bg_ptr,
+                                 <int> self.n_rows,
+                                 <int> n_cols,
+                                 <double*> row_ptr,
+                                 <int*> idx_ptr,
+                                 <bool> row_major)
+
+            self.handle.sync()
+
+            results = model_call(X=masked_inputs,
+                                 model=self.model,
+                                 model_gpu_based=self.model_gpu_based)
+
+            self.obs_output = results[len(results) / 2]
+
+            results = results.reshape(2 * n_cols + 1, len(self.masker))
+
+            averaged_outs = cp.mean(cp.asarray(self.link_fn(results)), axis=1)
+
+            tpb = 512
+            blks = int((len(idx) + tpb - 1) / tpb)
+
+            if dtype == cp.float32:
+                _row_values_kernel_f32((blks,), (tpb,), (row_values,
+                                                         averaged_outs,
+                                                         idx,
+                                                         len(idx)))
+            else:
+                _row_values_kernel_f64((blks,), (tpb,), (row_values,
+                                                         averaged_outs,
+                                                         idx,
+                                                         len(idx)))
+
+        row_values /= (2 * npermutations)
+        self.expected_value = averaged_outs[0]
+
+        diff = cp.sum(row_values) - (self.obs_output - self.expected_value)
+
+        if main_effects:
+            del masked_inputs
+            main_effect_values = self.main_effects(row)
+        else:
+            main_effect_values = None
+
+        return row_values, main_effect_values
+
+    def main_effects(self, row, inds=None):
+        if inds is None:
+            inds = cp.arange(len(self.masker), dtype=np.float32)
+
+        masked_inputs = cp.empty(
+            shape=((self.n_rows * self.n_cols + self.n_rows), self.n_cols),
+            dtype=self.dtype,
+            order=self.masker.order
+        )
+
+        cdef handle_t* handle_ = \
+            <handle_t*><size_t>self.handle.getHandle()
+        cdef uintptr_t row_ptr, bg_ptr, idx_ptr, masked_ptr
+
+        masked_ptr = masked_inputs.__cuda_array_interface__['data'][0]
+        bg_ptr = self.masker.ptr
+        row_ptr = row.ptr
+        idx_ptr = inds.__cuda_array_interface__['data'][0]
+        row_major = self.masker.order == "C"
+
+        if self.masker.order.dtype == cp.float32:
```

Comment:
why do we need the dtype from "order" instead of masker directly?

Reply:
typo that came from a wrong find and replace in Sublime